src/main/java/iabudiab/maven/plugins/dependencytrack/client/DTrackClient.java

```diff
@@ -193,13 +195,48 @@ public List<Finding> getProjectFindings(UUID projectId) throws IOException {
 		return Arrays.asList(findings);
 	}
 
+	public ProjectMetrics getProjectMetrics(UUID projectId, int retryDelay, int retryLimit) throws IOException {
+		if(retryLimit <= 0) return getProjectMetrics(projectId);
+		if(retryDelay < 0) throw new IllegalArgumentException("project metrics retry delay must be >= 0");
+
+		Supplier<ProjectMetrics> projectMetricsSupplier = () -> {
+			ProjectMetrics metrics = null;
+			try {
+				metrics = getProjectMetrics(projectId);
+			} catch(Exception ex) {
+				log.warn("got exception while obtaining project metrics for project '"+ projectId +"'", ex);
+				return null;
+			}
+			return metrics;
+		};
+		return retry(projectMetricsSupplier, metric -> metric == null, retryDelay, 0, retryLimit).join();
+	}
+
 	public ProjectMetrics getProjectMetrics(UUID projectId) throws IOException {
 		URI uri = baseUri.resolve(API_PROJECT_METRICS + projectId.toString() + "/current");
 		log.debug("Invoking uri => " + uri);
 		HttpGet request = httpGet(uri);
 		return client.execute(request, responseBodyHandler(ProjectMetrics.class));
 	}
 
+	public <R> CompletableFuture<R> retry(Supplier<R> supplier, Function<R, Boolean> retryCondition, int retryDelay, int retryCount, int retryLimit) {
```

Comment:
You are right, there is no reason for this method to be public. And sure, this method is implemented abstractly enough to be used for other retry or polling purposes.

Should we replace existing implementations (like polling Token processing) to use this method either or should we use it for new implementations in the future?

Reply:
I've moved the retry method to the `Utils` class since it is unspecific enough and so we can reuse it in other parts of the code, not only when contacting the `DTrackClient`. This way we can easily keep the unit tests, too.

What do you think about that?

Reply:
Yes, we could do that. It is somewhat similar to RetryTemplate in Spring or Retry in resilience4j. However, I'll change the API a little bit later 😉